cpp/src/text/subword/bpe_tokenizer.cu

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <text/subword/bpe_tokenizer.cuh>
+
+#include <nvtext/bpe_tokenize.hpp>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/combine.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/for_each.h>
+#include <thrust/merge.h>
+#include <thrust/transform.h>
+
+namespace nvtext {
+namespace detail {
+
+namespace {
+
+template <typename CharType>
+constexpr bool is_whitespace(CharType ch)
+{
+  return ch <= ' ';
+}
+
+/**
+ * @brief Resolve a truncated string from a full string.
+ *
+ * This will return a substring of the input starting with the first byte
+ * upto the first whitespace character is found or the end of the string.
+ * Any whitespace is expected only at the end of the string.
+ *
+ * @param d_str Input string to resolve.
+ * @return Substring of the input excluding trailing whitespace.
+ */
+__device__ cudf::string_view resolve_string(cudf::string_view const& d_str)
```

Comment:
I did not like `rstrip` or `trim` since the behavior here does not match what those functions perform. The [`std::substr`](https://en.cppreference.com/w/cpp/string/basic_string_view/substr) is closer in behavior but does not convey the operation here exactly.